### PR TITLE
assert: add `NotZeroThenSetZero` and `NotNilThenSetNil`

### PR DESCRIPTION
**Pull request by HaraldNordgren**
I have used similar versions in my own code for many years. It helps to do assertions of API responses, like this:

```go
clientResponse := someApiRequest(request)

assert.NotZeroThenSetZero(t, &clientResponse.CreateTime)
assert.NotZeroThenSetZero(t, &clientResponse.UpdateTime)

assert.Equal(t, expected, clientResponse)
```

Could resolve the issue brought up in https://github.com/stretchr/testify/issues/1432.

### CODE REVIEW

**Comment by dolmen** [comment]
1. I don't understand the use case. If this API usage is not obvious, then it's place is probably not the core testify.
2. Implantation requires generics. This is a blocking for now as we still aim to support Go versions below 1.18.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -24,7 +24,6 @@ jobs:
     strategy:
       matrix:
         go_version:
-          - "1.17"
```
Comment (dolmen) [inline .github/workflows/main.yml:27]: Don't change this. This is out of scope.

File: assert/assertions.go [hunk 2]
```diff
@@ -620,6 +620,39 @@ func NotNil(t TestingT, object interface{}, msgAndArgs ...interface{}) bool {
...
 
+// NotNilThenSetNil asserts that the specified object is not zero and then sets it to zero.
+//
+//	assert.NotZeroThenSetZero(t, &x)
+func NotZeroThenSetZero[X any](t TestingT, x *X) bool {
+	if h, ok := t.(tHelper); ok {
+		h.Helper()
+	}
+
+	if ok := NotZero(t, *x); !ok {
```
Comment (dolmen) [inline assert/assertions.go:631]: The ok variable can be removed.

File: assert/assertions.go [hunk 3]
```diff
@@ -620,6 +620,39 @@ func NotNil(t TestingT, object interface{}, msgAndArgs ...interface{}) bool {
...
+
+	var zeroValue X
+	*x = zeroValue
+	return true
+}
+
+// NotNilThenSetNil asserts that the specified pointer is not nil and then sets it to nil.
+//
+//	assert.NotNilThenSetNil(t, &x)
+func NotNilThenSetNil[X any](t TestingT, x **X) bool {
```
Comment (dolmen) [inline assert/assertions.go:643]: This func is redundant with NotZeroThenSetZero.

1. Why 2 funcs?
2. Why duplicate the implementation instead of wrapping one with the other?

**Review by dolmen — CHANGES_REQUESTED** [review]
I do not plan to merge this.

But anyway I choose to give you feedback that might be valuable for another contribution 

Also you didn't run "go generate ./.. ".

File: .github/workflows/main.yml (hunk 1 above)
Comment (HaraldNordgren) [reply .github/workflows/main.yml:27]: I changed the minimum version to be >=1.18 in discussion with Boyan Soubachov in this PR: https://github.com/stretchr/testify/pull/1379#discussion_r1188456518. Maybe that was a bad change?

**Comment by HaraldNordgren** [comment]
I'm sorry you you feel you don't want to merge this.

> Also you didn't run "go generate ./.. ".

Yes, I noticed that the cannot run now. Seems to be because of generics as well. Should I close the PR?

**Comment by dolmen** [comment]
From your description of the use case:

```go
clientResponse := someApiRequest(request)

assert.NotZeroThenSetZero(t, &clientResponse.CreateTime)
assert.NotZeroThenSetZero(t, &clientResponse.UpdateTime)

assert.Equal(t, expected, clientResponse)
```

This can be rewritten as:
```go
clientResponse := someApiRequest(request)

assert.NotZero(t, clientResponse.CreateTime)
assert.NotZero(t, clientResponse.UpdateTime)
var zeroTime time.Time
clientResponse.CreateTime = zeroTime
clientResponse.UpdateTime = zeroTime

assert.Equal(t, expected, clientResponse)
```

And I think that is would be more clear that the proposed API.

I think that if you want than feature it is best to copy your implementation in the test package where you use it. That way, the source is immediately available to the reader who may not be familiar with the full `testify` API. We don't need more niche features.

**Comment by dolmen** [comment]
About the code generation not working with generics code, this is a problem that we must fix before introducing any generics-based API.
